### PR TITLE
Remove property-based schema comparison APIs

### PR DESCRIPTION
**Pull request by morozov**
**TODO**:
- [x] https://github.com/doctrine/dbal/pull/5650

### CODE REVIEW

File: tests/Platforms/AbstractPlatformTestCase.php [hunk 1]
```diff
@@ -239,35 +239,20 @@ public function testGeneratesTableAlterationSql(): void
...
         $table->setPrimaryKey(['id']);
 
-        $tableDiff                        = new TableDiff('mytable');
-        $tableDiff->fromTable             = $table;
-        $tableDiff->newName               = 'userlist';
-        $tableDiff->addedColumns['quota'] = new Column('quota', Type::getType('integer'), ['notnull' => false]);
-        $tableDiff->removedColumns['foo'] = new Column('foo', Type::getType('integer'));
-        $tableDiff->changedColumns['bar'] = new ColumnDiff(
-            new Column(
-                'baz',
```
Comment (morozov) [inline tests/Platforms/AbstractPlatformTestCase.php:253]: This part of the test is invalid.

First of all, the column in `$changedColumns` cannot have its name changed. The old and the new column should have the same name to be added to `$changedColumns`: 
https://github.com/doctrine/dbal/blob/58d4b3c78988dec39294a4508cd3272265e065db/src/Schema/Comparator.php#L205-L225

Second, the test expects a rename column SQL to be generated for MySQL: https://github.com/doctrine/dbal/blob/08630c6188d70b0f7a98b9c4a079f5f7a76b6cba/tests/Platforms/AbstractMySQLPlatformTestCase.php#L64

But not on Oracle: https://github.com/doctrine/dbal/blob/4bb2c0b2bf735639b1e32796e506f8be9fcac7f7/tests/Platforms/OraclePlatformTest.php#L53

Note that the latter assumes that the column name is already "baz", not "bar".

File: tests/Platforms/SQLServerPlatformTestCase.php [hunk 2]
```diff
@@ -50,29 +50,6 @@ public function getGenerateTableWithMultiColumnUniqueIndexSql(): array
...
-            'ALTER TABLE mytable ALTER COLUMN bloo BIT NOT NULL',
-            'ALTER TABLE mytable ADD CONSTRAINT DF_6B2BD609_CECED971 DEFAULT 0 FOR bloo',
-            "sp_rename 'mytable', 'userlist'",
-            "DECLARE @sql NVARCHAR(MAX) = N''; " .
-            "SELECT @sql += N'EXEC sp_rename N''' + dc.name + ''', N''' " .
-            "+ REPLACE(dc.name, '6B2BD609', 'E2B58069') + ''', ''json'';' " .
-            'FROM sys.default_constraints dc ' .
-            'JOIN sys.tables tbl ON dc.parent_object_id = tbl.object_id ' .
-            "WHERE tbl.name = 'userlist';" .
-            'EXEC sp_executesql @sql',
```
Comment (morozov) [inline tests/Platforms/SQLServerPlatformTestCase.php:72]: Assertions where the expected SQL was copy-pasted from the debugger don't add any value. The only thing they prove is that the generated SQL will not accidentally change which complicates changing code (including fixing bugs and refactoring). As the above comment shows, the expected SQL is not always correct.

File: tests/Schema/ComparatorTest.php [hunk 3]
```diff
@@ -818,17 +825,6 @@ public function testChangedSequence(): void
...
-    /** @psalm-suppress NullArgument */
-    public function testDiffDecimalWithNullPrecision(): void
-    {
-        $column = new Column('foo', Type::getType('decimal'));
-        $column->setPrecision(null);
-
-        $column2 = new Column('foo', Type::getType('decimal'));
-
-        self::assertEquals([], $this->comparator->diffColumn($column, $column2));
-    }
```
Comment (morozov) [inline tests/Schema/ComparatorTest.php:814]: This test only tests that the default precision is null. It is pointless.

File: tests/Schema/ComparatorTest.php [hunk 4]
```diff
@@ -1130,34 +1091,17 @@ public function testComparesNamespaces(): void
...
-        $column1 = new Column('foo', Type::getType('guid'), ['comment' => 'GUID 1']);
-        $column2 = new Column(
-            'foo',
-            Type::getType('guid'),
-            ['notnull' => false, 'length' => 36, 'fixed' => true, 'default' => 'NEWID()', 'comment' => 'GUID 2.'],
-        );
-
-        self::assertEquals(['notnull', 'default', 'comment'], $this->comparator->diffColumn($column1, $column2));
-        self::assertEquals(['notnull', 'default', 'comment'], $this->comparator->diffColumn($column2, $column1));
-    }
```
Comment (morozov) [inline tests/Schema/ComparatorTest.php:1118]: The comparator no longer needs to be aware of platform-specific type implementation details. This is taken care of in the platform-aware schema comparison.

File: tests/Schema/ComparatorTest.php [hunk 5]
```diff
@@ -1064,39 +1058,6 @@ public function assertSchemaSequenceChangeCount(
         self::assertCount($removeSequenceCount, $diff->removedSequences);
     }
 
-    public function testDiffColumnPlatformOptions(): void
```
Comment (morozov) [inline tests/Schema/ComparatorTest.php:1041]: Changes in platform options are no longer exposed via the column diff as they were never used anywhere except for this test.

File: src/Schema/ColumnDiff.php [hunk 6]
```diff
@@ -4,87 +4,95 @@
...
 
     public function hasDefaultChanged(): bool
     {
-        return $this->hasChanged('default');
+        $oldDefault = $this->fromColumn->getDefault();
+        $newDefault = $this->column->getDefault();
+
+        // Null values need to be checked additionally as they tell whether to create or drop a default value.
+        // null != 0, null != false, null != '' etc. This affects platform's table alteration SQL generation.
+        if (($newDefault === null) !== ($oldDefault === null)) {
```
Comment (derrabus) [inline src/Schema/ColumnDiff.php]:
```suggestion
        if (($newDefault === null) xor ($oldDefault === null)) {
```

Maybe it's a me-only problem but `!==` on booleans makes my head spin. 😓

Comment (morozov) [reply src/Schema/ColumnDiff.php]: I'm curious what it would look like if we wanted to check them for equality 🤔

**Review by derrabus — APPROVED** [review]

File: src/Schema/ColumnDiff.php (hunk 6 above)
Comment (derrabus) [reply src/Schema/ColumnDiff.php]: Let's start an RFC to add the `xnor` operator to PHP. 😁

**Pull request merged**